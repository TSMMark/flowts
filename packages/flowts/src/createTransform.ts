import * as babel from '@babel/core';
import * as prettier from 'prettier';
import recastPlugin from '@flowts/babel-plugin-recast';
import { PluginItem } from '@babel/core';
import { Options } from 'prettier';

/**
 * Helper utility creating function transforming the code given plugins array
 * @param plugins
 * @param prettierConfig
 * @param recast
 */
export function createTransform(
  plugins: PluginItem[],
  { prettierConfig = {} as Options, recast = true } = {}
) {
  const transformPlugins: PluginItem[] = [];
  if (recast) {
    transformPlugins.push(recastPlugin);
  }

  return function transform(code: string) {
    const transformResult = babel.transformSync(code, {
      compact: false,
      babelrc: false,
      configFile: false,
      plugins: [...transformPlugins, ...plugins],
    });

    if (transformResult === null) {
      throw new Error(
        'babel.transformSync returned null, probably configuration issue'
      );
    }
    prettierConfig.parser = 'babel-ts';
    return prettier.format(transformResult.code as string, {
      ...prettierConfig,
      // avoid changing template strings
      embeddedLanguageFormatting: 'off',
    });
  };
}
