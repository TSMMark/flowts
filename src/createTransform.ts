import * as babel from '@babel/core';
import * as prettier from 'prettier';
import recastPlugin from './recastPlugin';
import { PluginItem } from '@babel/core';
import { Options } from 'prettier';

/**
 * Helper utility creating function transforming the code given plugins array
 * @param plugins
 * @param prettierConfig
 * @param recast
 */
export function createTransform(
  plugins: PluginItem[],
  { prettierConfig = {} as Options, recast = true } = {}
) {
  const transformPlugins: PluginItem[] = [];
  if (recast) {
    transformPlugins.push(recastPlugin);
  }

  return function transform(code: string) {
    const transformResult = babel.transformSync(code, {
      babelrc: false,
      plugins: [...transformPlugins, ...plugins],
    });

    if (transformResult === null) {
      throw new Error(
        'babel.transformSync returned null, probably configuration issue'
      );
    }
    prettierConfig.parser = 'typescript';
    return prettier.format(transformResult.code as string, prettierConfig);
  };
}
