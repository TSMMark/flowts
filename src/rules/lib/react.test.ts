import plugin from '../../tsTypesPlugin';
import { createTransform } from '../../createTransform';
const transform = createTransform([plugin]);

describe('lib/react.js', () => {
  describe('globals', () => {
    describe('React$Node', () => {
      test('generated - type', () => {
        expect(
          transform(`
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface', () => {
        expect(
          transform(`
            class A1 extends React$Node {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('React$Component', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = React$Component;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class - typePrams(1 - 2)', () => {
        expect(
          transform(`
            new React$Component();
            
            class A1 extends React$Component<P0> {}
            let a1 = new React$Component<P0>();
            
            class A2 extends React$Component<P0, P1> {}
            let a2 = new React$Component<P0, P1>();
            
          `)
        ).toMatchSnapshot();
      });
    });

    describe('React$PureComponent', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = React$PureComponent;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class - typePrams(1 - 2)', () => {
        expect(
          transform(`
            new React$PureComponent();
            
            class A1 extends React$PureComponent<P0> {}
            let a1 = new React$PureComponent<P0>();
            
            class A2 extends React$PureComponent<P0, P1> {}
            let a2 = new React$PureComponent<P0, P1>();
            
          `)
        ).toMatchSnapshot();
      });
    });

    describe('LegacyReactComponent', () => {
      test('variable', () => {
        expect(
          transform(`
            var a = LegacyReactComponent;
          `)
        ).toMatchSnapshot();
      });

      test('generated - class - typePrams(2)', () => {
        expect(
          transform(`
            new LegacyReactComponent();
            
            class A1 extends LegacyReactComponent<P0, P1> {}
            let a1 = new LegacyReactComponent<P0, P1>();
            
          `)
        ).toMatchSnapshot();
      });
    });

    describe('React$AbstractComponentStatics', () => {
      test('generated - type', () => {
        expect(
          transform(`
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface', () => {
        expect(
          transform(`
            class A1 extends React$AbstractComponentStatics {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('React$StatelessFunctionalComponent', () => {
      test('generated - type - typePrams(1)', () => {
        expect(
          transform(`
            let a1: React$StatelessFunctionalComponent<P0>;
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface - typePrams(1)', () => {
        expect(
          transform(`
            class A1 implements React$StatelessFunctionalComponent<P0> {}
            interface I1 extends React$StatelessFunctionalComponent<P0> {}
            
          `)
        ).toMatchSnapshot();
      });
    });

    describe('React$ComponentType', () => {
      test('generated - type - typePrams(1)', () => {
        expect(
          transform(`
            let a1: React$ComponentType<P0>;
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface - typePrams(1)', () => {
        expect(
          transform(`
            class A1 implements React$ComponentType<P0> {}
            interface I1 extends React$ComponentType<P0> {}
            
          `)
        ).toMatchSnapshot();
      });
    });

    describe('React$ElementType', () => {
      test('generated - type', () => {
        expect(
          transform(`
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface', () => {
        expect(
          transform(`
            class A1 extends React$ElementType {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('React$Element', () => {
      test('generated - type - typePrams(1)', () => {
        expect(
          transform(`
            let a1: React$Element<P0>;
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface - typePrams(1)', () => {
        expect(
          transform(`
            class A1 implements React$Element<P0> {}
            interface I1 extends React$Element<P0> {}
            
          `)
        ).toMatchSnapshot();
      });
    });

    describe('React$MixedElement', () => {
      test('generated - type', () => {
        expect(
          transform(`
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface', () => {
        expect(
          transform(`
            class A1 extends React$MixedElement {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('React$Key', () => {
      test('generated - type', () => {
        expect(
          transform(`
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface', () => {
        expect(
          transform(`
            class A1 extends React$Key {};
          `)
        ).toMatchSnapshot();
      });
    });

    describe('React$Ref', () => {
      test('generated - type - typePrams(1)', () => {
        expect(
          transform(`
            let a1: React$Ref<P0>;
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface - typePrams(1)', () => {
        expect(
          transform(`
            class A1 implements React$Ref<P0> {}
            interface I1 extends React$Ref<P0> {}
            
          `)
        ).toMatchSnapshot();
      });
    });

    describe('React$Context', () => {
      test('generated - type - typePrams(1)', () => {
        expect(
          transform(`
            let a1: React$Context<P0>;
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface - typePrams(1)', () => {
        expect(
          transform(`
            class A1 implements React$Context<P0> {}
            interface I1 extends React$Context<P0> {}
            
          `)
        ).toMatchSnapshot();
      });
    });

    describe('React$Portal', () => {
      test('generated - type', () => {
        expect(
          transform(`
          `)
        ).toMatchSnapshot();
      });

      test('generated - interface', () => {
        expect(
          transform(`
            class A1 extends React$Portal {};
          `)
        ).toMatchSnapshot();
      });
    });
  });

  describe('modules', () => {
    describe('react', () => {
      describe('DOM', () => {
        test('has no test', () => {});
      });

      describe('PropTypes', () => {
        test('has no test', () => {});
      });

      describe('version', () => {
        test('has no test', () => {});
      });

      describe('checkPropTypes', () => {
        test('has no test', () => {});
      });

      describe('createClass', () => {
        test('has no test', () => {});
      });

      describe('createContext', () => {
        test('has no test', () => {});
      });

      describe('createElement', () => {
        test('has no test', () => {});
      });

      describe('cloneElement', () => {
        test('has no test', () => {});
      });

      describe('createFactory', () => {
        test('has no test', () => {});
      });

      describe('createRef', () => {
        test('has no test', () => {});
      });

      describe('isValidElement', () => {
        test('has no test', () => {});
      });

      describe('Component', () => {
        test('has no test', () => {});
      });

      describe('PureComponent', () => {
        test('has no test', () => {});
      });

      describe('StatelessFunctionalComponent', () => {
        test('has no test', () => {});
      });

      describe('ComponentType', () => {
        test('has no test', () => {});
      });

      describe('AbstractComponent', () => {
        test('has no test', () => {});
      });

      describe('MixedElement', () => {
        test('has no test', () => {});
      });

      describe('ElementType', () => {
        test('has no test', () => {});
      });

      describe('Element', () => {
        test('has no test', () => {});
      });

      describe('Fragment', () => {
        test('has no test', () => {});
      });

      describe('Key', () => {
        test('has no test', () => {});
      });

      describe('Ref', () => {
        test('has no test', () => {});
      });

      describe('Node', () => {
        test('has no test', () => {});
      });

      describe('Context', () => {
        test('has no test', () => {});
      });

      describe('Portal', () => {
        test('has no test', () => {});
      });

      describe('ConcurrentMode', () => {
        test('has no test', () => {});
      });

      describe('StrictMode', () => {
        test('has no test', () => {});
      });

      describe('Suspense', () => {
        test('has no test', () => {});
      });

      describe('ElementProps', () => {
        test('has no test', () => {});
      });

      describe('ElementConfig', () => {
        test('has no test', () => {});
      });

      describe('ElementRef', () => {
        test('has no test', () => {});
      });

      describe('Config', () => {
        test('has no test', () => {});
      });

      describe('ChildrenArray', () => {
        test('has no test', () => {});
      });

      describe('Children', () => {
        test('has no test', () => {});
      });

      describe('forwardRef', () => {
        test('has no test', () => {});
      });

      describe('memo', () => {
        test('has no test', () => {});
      });

      describe('lazy', () => {
        test('has no test', () => {});
      });

      describe('MaybeCleanUpFn', () => {
        test('generated - type', () => {
          expect(
            transform(`
            import { MaybeCleanUpFn } from "react";
            
          `)
          ).toMatchSnapshot();
        });

        test('generated - interface', () => {
          expect(
            transform(`
            import { MaybeCleanUpFn } from "react";
            
            class A1 extends MaybeCleanUpFn {};
          `)
          ).toMatchSnapshot();
        });

        test('generated - type - import namespace', () => {
          expect(
            transform(`
            import * as M from "react";
            
          `)
          ).toMatchSnapshot();
        });

        test('generated - interface - import namespace', () => {
          expect(
            transform(`
            import * as M from "react";
            
            class A1 extends M.MaybeCleanUpFn {};
          `)
          ).toMatchSnapshot();
        });

        test('generated - type - import renamed', () => {
          expect(
            transform(`
            import { MaybeCleanUpFn as t } from "react";
            
          `)
          ).toMatchSnapshot();
        });

        test('generated - interface - import renamed', () => {
          expect(
            transform(`
            import { MaybeCleanUpFn as t } from "react";
            
            class A1 extends t {};
          `)
          ).toMatchSnapshot();
        });
      });

      describe('useContext', () => {
        test('has no test', () => {});
      });

      describe('useState', () => {
        test('has no test', () => {});
      });

      describe('Dispatch', () => {
        test('generated - type - typePrams(1)', () => {
          expect(
            transform(`
            import { Dispatch } from "react";
            
            let a1: Dispatch<P0>;
          `)
          ).toMatchSnapshot();
        });

        test('generated - interface - typePrams(1)', () => {
          expect(
            transform(`
            import { Dispatch } from "react";
            
            class A1 implements Dispatch<P0> {}
            interface I1 extends Dispatch<P0> {}
            
          `)
          ).toMatchSnapshot();
        });

        test('generated - type - import namespace - typePrams(1)', () => {
          expect(
            transform(`
            import * as M from "react";
            
            let a1: M.Dispatch<P0>;
          `)
          ).toMatchSnapshot();
        });

        test('generated - interface - import namespace - typePrams(1)', () => {
          expect(
            transform(`
            import * as M from "react";
            
            class A1 implements M.Dispatch<P0> {}
            interface I1 extends M.Dispatch<P0> {}
            
          `)
          ).toMatchSnapshot();
        });

        test('generated - type - import renamed - typePrams(1)', () => {
          expect(
            transform(`
            import { Dispatch as t } from "react";
            
            let a1: t<P0>;
          `)
          ).toMatchSnapshot();
        });

        test('generated - interface - import renamed - typePrams(1)', () => {
          expect(
            transform(`
            import { Dispatch as t } from "react";
            
            class A1 implements t<P0> {}
            interface I1 extends t<P0> {}
            
          `)
          ).toMatchSnapshot();
        });
      });

      describe('useReducer', () => {
        test('has no test', () => {});
      });

      describe('useRef', () => {
        test('has no test', () => {});
      });

      describe('useDebugValue', () => {
        test('has no test', () => {});
      });

      describe('useEffect', () => {
        test('has no test', () => {});
      });

      describe('useLayoutEffect', () => {
        test('has no test', () => {});
      });

      describe('useCallback', () => {
        test('has no test', () => {});
      });

      describe('useMemo', () => {
        test('has no test', () => {});
      });

      describe('useImperativeHandle', () => {
        test('has no test', () => {});
      });

      describe('default', () => {
        test('has no test', () => {});
      });
    });

    describe('React', () => {
      describe('default', () => {
        test('has no test', () => {});
      });
    });
  });
});
