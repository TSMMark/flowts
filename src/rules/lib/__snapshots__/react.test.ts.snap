// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/react.js globals LegacyReactComponent generated - class - typePrams(2) 1`] = `
"new LegacyReactComponent();

class A1 extends LegacyReactComponent<P0, P1> {}
let a1 = new LegacyReactComponent<P0, P1>();
"
`;

exports[`lib/react.js globals LegacyReactComponent variable 1`] = `
"var a = LegacyReactComponent;
"
`;

exports[`lib/react.js globals React$AbstractComponentStatics generated - interface 1`] = `
"class A1 extends React$AbstractComponentStatics {}
"
`;

exports[`lib/react.js globals React$AbstractComponentStatics generated - type 1`] = `""`;

exports[`lib/react.js globals React$Component generated - class - typePrams(1 - 2) 1`] = `
"import { Component } from \\"react\\";
new Component();

class A1 extends Component<P0> {}
let a1 = new Component<P0>();

class A2 extends Component<P0, P1> {}
let a2 = new Component<P0, P1>();
"
`;

exports[`lib/react.js globals React$Component variable 1`] = `
"import { Component } from \\"react\\";
var a = Component;
"
`;

exports[`lib/react.js globals React$ComponentType generated - interface - typePrams(1) 1`] = `
"import { ComponentType } from \\"react\\";
class A1 implements ComponentType<P0> {}
interface I1 extends ComponentType<P0> {}
"
`;

exports[`lib/react.js globals React$ComponentType generated - type - typePrams(1) 1`] = `
"import { ComponentType } from \\"react\\";
let a1: ComponentType<P0>;
"
`;

exports[`lib/react.js globals React$Context generated - interface - typePrams(1) 1`] = `
"import { Context } from \\"react\\";
class A1 implements Context<P0> {}
interface I1 extends Context<P0> {}
"
`;

exports[`lib/react.js globals React$Context generated - type - typePrams(1) 1`] = `
"import { Context } from \\"react\\";
let a1: Context<P0>;
"
`;

exports[`lib/react.js globals React$Element generated - interface - typePrams(1) 1`] = `
"import { Element } from \\"react\\";
class A1 implements Element<P0> {}
interface I1 extends Element<P0> {}
"
`;

exports[`lib/react.js globals React$Element generated - type - typePrams(1) 1`] = `
"import { Element } from \\"react\\";
let a1: Element<P0>;
"
`;

exports[`lib/react.js globals React$ElementType generated - interface 1`] = `
"import { ElementType } from \\"react\\";
class A1 extends ElementType {}
"
`;

exports[`lib/react.js globals React$ElementType generated - type 1`] = `""`;

exports[`lib/react.js globals React$Key generated - interface 1`] = `
"import { Key } from \\"react\\";
class A1 extends Key {}
"
`;

exports[`lib/react.js globals React$Key generated - type 1`] = `""`;

exports[`lib/react.js globals React$MixedElement generated - interface 1`] = `
"import { MixedElement } from \\"react\\";
class A1 extends MixedElement {}
"
`;

exports[`lib/react.js globals React$MixedElement generated - type 1`] = `""`;

exports[`lib/react.js globals React$Node generated - interface 1`] = `
"import { Node } from \\"react\\";
class A1 extends Node {}
"
`;

exports[`lib/react.js globals React$Node generated - type 1`] = `""`;

exports[`lib/react.js globals React$Portal generated - interface 1`] = `
"import { Portal } from \\"react\\";
class A1 extends Portal {}
"
`;

exports[`lib/react.js globals React$Portal generated - type 1`] = `""`;

exports[`lib/react.js globals React$PureComponent generated - class - typePrams(1 - 2) 1`] = `
"import { PureComponent } from \\"react\\";
new PureComponent();

class A1 extends PureComponent<P0> {}
let a1 = new PureComponent<P0>();

class A2 extends PureComponent<P0, P1> {}
let a2 = new PureComponent<P0, P1>();
"
`;

exports[`lib/react.js globals React$PureComponent variable 1`] = `
"import { PureComponent } from \\"react\\";
var a = PureComponent;
"
`;

exports[`lib/react.js globals React$Ref generated - interface - typePrams(1) 1`] = `
"import { Ref } from \\"react\\";
class A1 implements Ref<P0> {}
interface I1 extends Ref<P0> {}
"
`;

exports[`lib/react.js globals React$Ref generated - type - typePrams(1) 1`] = `
"import { Ref } from \\"react\\";
let a1: Ref<P0>;
"
`;

exports[`lib/react.js globals React$StatelessFunctionalComponent generated - interface - typePrams(1) 1`] = `
"import { StatelessFunctionalComponent } from \\"react\\";
class A1 implements StatelessFunctionalComponent<P0> {}
interface I1 extends StatelessFunctionalComponent<P0> {}
"
`;

exports[`lib/react.js globals React$StatelessFunctionalComponent generated - type - typePrams(1) 1`] = `
"import { StatelessFunctionalComponent } from \\"react\\";
let a1: StatelessFunctionalComponent<P0>;
"
`;

exports[`lib/react.js modules react Dispatch generated - interface - import namespace - typePrams(1) 1`] = `
"import * as M from \\"react\\";

class A1 implements M.Dispatch<P0> {}
interface I1 extends M.Dispatch<P0> {}
"
`;

exports[`lib/react.js modules react Dispatch generated - interface - import renamed - typePrams(1) 1`] = `
"import { Dispatch as t } from \\"react\\";

class A1 implements t<P0> {}
interface I1 extends t<P0> {}
"
`;

exports[`lib/react.js modules react Dispatch generated - interface - typePrams(1) 1`] = `
"import { Dispatch } from \\"react\\";

class A1 implements Dispatch<P0> {}
interface I1 extends Dispatch<P0> {}
"
`;

exports[`lib/react.js modules react Dispatch generated - type - import namespace - typePrams(1) 1`] = `
"import * as M from \\"react\\";

let a1: M.Dispatch<P0>;
"
`;

exports[`lib/react.js modules react Dispatch generated - type - import renamed - typePrams(1) 1`] = `
"import { Dispatch as t } from \\"react\\";

let a1: t<P0>;
"
`;

exports[`lib/react.js modules react Dispatch generated - type - typePrams(1) 1`] = `
"import { Dispatch } from \\"react\\";

let a1: Dispatch<P0>;
"
`;

exports[`lib/react.js modules react MaybeCleanUpFn generated - interface - import namespace 1`] = `
"import * as M from \\"react\\";

class A1 extends M.MaybeCleanUpFn {}
"
`;

exports[`lib/react.js modules react MaybeCleanUpFn generated - interface - import renamed 1`] = `
"import { MaybeCleanUpFn as t } from \\"react\\";

class A1 extends t {}
"
`;

exports[`lib/react.js modules react MaybeCleanUpFn generated - interface 1`] = `
"import { MaybeCleanUpFn } from \\"react\\";

class A1 extends MaybeCleanUpFn {}
"
`;

exports[`lib/react.js modules react MaybeCleanUpFn generated - type - import namespace 1`] = `
"import * as M from \\"react\\";
"
`;

exports[`lib/react.js modules react MaybeCleanUpFn generated - type - import renamed 1`] = `
"import { MaybeCleanUpFn as t } from \\"react\\";
"
`;

exports[`lib/react.js modules react MaybeCleanUpFn generated - type 1`] = `
"import { MaybeCleanUpFn } from \\"react\\";
"
`;
